### PR TITLE
Add initial version of HdrHistogram

### LINKED ISSUE #39: Basic C++ HdrHistogram implementation
To avoid external dependencies, implement a pure C++ version of HdrHistogram.

The current goal includes only a subset of the features supported in the original version.

### PR DESCRIPTION
**Pull request by rfernandes**
Integration with bencharking utils via -o hdr and -o stat options

While functional from missing things include:
  * Correct iterator semantics
  * Load / Dump capabilities
  * Support for different sizes (currently using uint64_t)

Close #39

### CODE REVIEW

File: toolbox/util/HdrHistogram.cpp [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+}
+
+using RecordedIteratorAdapter = IteratorAdapter<RecordedIterator>;
+
+double HdrHistogram::mean() const
+{
+    if (!total_count_) {
+        return 0.0;
+    }
+    double total{0.0};
```
Comment (marayl) [inline toolbox/util/HdrHistogram.cpp]: The two variables below are unused. Names should also be snake_case.

File: toolbox/util/HdrHistogram.cpp [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "HdrHistogram.hpp"
+
+namespace {
+std::size_t get_bucket_count(uint64_t value, std::size_t subb_count, std::size_t unit_mag)
```
Comment (marayl) [inline toolbox/util/HdrHistogram.cpp]: `noexcept`

File: toolbox/util/HdrHistogram.cpp [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+    //         if (index >= encoder.payload.counts_len_){
+    //             return 0;
+    //         }
+    return counts_.at(index);
+}
+
+uint64_t HdrHistogram::get_target_count_at_percentile(double percentile)
+{
+    auto requested_percentile = std::min(percentile, 100.0);
+    int count_at_percentile = ((requested_percentile * total_count_ / 100)) + 0.5;
```
Comment (marayl) [inline toolbox/util/HdrHistogram.cpp]: Why double parens?

File: toolbox/util/HdrHistogram.hpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+#include <toolbox/Config.h>
+
+#include <cmath>
+#include <iomanip>
+#include <iostream>
+#include <limits>
+#include <vector>
+
+namespace toolbox {
```
Comment (marayl) [inline toolbox/util/HdrHistogram.hpp]: Suggestion: move hdr components into hdr subdirectory and namespace.

File: toolbox/util/HdrHistogram.hpp [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+inline namespace util {
+
+class TOOLBOX_API HdrHistogram {
+
+  public:
+    HdrHistogram(std::size_t lowest_value, std::size_t highest_value,
```
Comment (marayl) [inline toolbox/util/HdrHistogram.hpp]: Use same types as HdrHistogram_c: signed `std::int64_t` for values.

File: toolbox/util/HdrHistogram.hpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    uint64_t get_lowest_equivalent_value(uint64_t value) const;
+    uint64_t get_value_from_sub_bucket(uint64_t bucket_index, uint64_t sub_bucket_index) const;
+    uint64_t get_count_at_index(std::size_t index) const;
+    uint64_t get_target_count_at_percentile(double percentile);
+    std::size_t counts_index_for(uint64_t value) const;
+
+    std::size_t get_bucket_index(uint64_t value) const;
+    std::size_t get_sub_bucket_index(uint64_t value, std::size_t bucket_index) const;
+
+    std::size_t clz(uint64_t value) const;
```
Comment (marayl) [inline toolbox/util/HdrHistogram.hpp]: Quite a few member functions could be `noexcept`.

File: toolbox/util/HdrHistogram.hpp [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    /// Returns:
+    ///    true if the two values are equivalentv
+    uint64_t values_are_equivalent(uint64_t lhs, uint64_t rhs);
+
+    uint64_t max() const noexcept;
+    uint64_t min() const;
+    double mean() const;
+    double stddev() const;
+
+    uint64_t get_count_at_value(uint64_t value);
```
Comment (marayl) [inline toolbox/util/HdrHistogram.hpp]: Make this `const noexcept`.

File: toolbox/util/HdrHistogram.hpp [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    uint64_t get_total_count() const;
+
+    uint64_t unit_magnitude_;
```
Comment (marayl) [inline toolbox/util/HdrHistogram.hpp]: No need for trailing underscore on public data-members. If these are intended to be private, then move to private section below.

File: toolbox/util/HdrHistogram.hpp [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+    uint64_t get_value_from_index(std::size_t index) const;
+    std::size_t counts_index(int bucket_index, int sub_bucket_index) const;
+
+    uint64_t hdr_size_of_equiv_value_range(uint64_t value) const
+    {
+        auto bucket_index = get_bucket_index(value);
+        auto sub_bucket_index = get_sub_bucket_index(value, bucket_index);
+        if (sub_bucket_index >= sub_bucket_count_) {
+            bucket_index += 1;
```
Comment (marayl) [inline toolbox/util/HdrHistogram.hpp]: Why not increment operator? Python port I suspect. There's at least one other instance of this.

File: toolbox/util/HdrHistogram.ut.cpp [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+#include <boost/test/unit_test.hpp>
+
+using namespace std;
+using namespace toolbox;
+
+namespace utf = boost::unit_test;
+
+BOOST_AUTO_TEST_SUITE(HdrHistogramSuite)
+
+constexpr auto LOWEST = 1;
```
Comment (marayl) [inline toolbox/util/HdrHistogram.ut.cpp]: Constant names as per style guide.

File: toolbox/util/HdrHistogramIterator.hpp [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+namespace toolbox {
+
+inline namespace util {
+
+class HdrHistogram;
+class AbstractHdrIterator;
+
+class HdrIterationValue {
+    const AbstractHdrIterator& iterator_;
```
Comment (marayl) [inline toolbox/util/HdrHistogramIterator.hpp]: Private data-members at the bottom on the class.

File: toolbox/util/HdrHistogram.hpp (hunk 4 above)
Comment (rfernandes) [reply toolbox/util/HdrHistogram.hpp]: Agree

File: toolbox/util/HdrHistogram.cpp (hunk 1 above)
Comment (rfernandes) [reply toolbox/util/HdrHistogram.cpp]: Changed iterAdapter casing, but they are used to iterate, unless I missed something ?

File: toolbox/util/HdrHistogram.hpp (hunk 4 above)
Comment (marayl) [reply toolbox/util/HdrHistogram.hpp]: I'd personally prefer to keep it flat and go with `toolbox/hdr`.

File: toolbox/util/HdrHistogram.cpp (hunk 1 above)
Comment (marayl) [reply toolbox/util/HdrHistogram.cpp]: Ah, yes. This comment was meant for the `stddev` function.

**Comment by marayl** [comment]
As always, please let me know if you need any help with the cosmetic cleanups.

File: toolbox/util/HdrHistogram.hpp (hunk 8 above)
Comment (rfernandes) [reply toolbox/util/HdrHistogram.hpp]: Made public members const

File: toolbox/util/HdrHistogram.hpp (hunk 7 above)
Comment (rfernandes) [reply toolbox/util/HdrHistogram.hpp]: Plenty other getters are now const noexcept

File: toolbox/util/HdrHistogram.hpp (hunk 4 above)
Comment (rfernandes) [reply toolbox/util/HdrHistogram.hpp]: Done

File: toolbox/util/HdrHistogram.cpp (hunk 3 above)
Comment (rfernandes) [reply toolbox/util/HdrHistogram.cpp]: Fixed

File: toolbox/util/HdrHistogram.hpp (hunk 5 above)
Comment (rfernandes) [reply toolbox/util/HdrHistogram.hpp]: Missed this one, will update later

File: toolbox/util/HdrHistogram.hpp (hunk 6 above)
Comment (rfernandes) [reply toolbox/util/HdrHistogram.hpp]: I believe I got all of them

File: toolbox/util/HdrHistogram.hpp (hunk 5 above)
Comment (rfernandes) [reply toolbox/util/HdrHistogram.hpp]: I left only the ones to do with indexes as std::size_t

File: toolbox/hdr/HdrHistogram.cpp [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+    // TODO: Move to make_iterator functor
+    const double output_value_unit_scaling_ratio = 1000.0;
+    out << "       Value     Percentile TotalCount 1/(1-Percentile)\n\n";
+    for (const auto& iter_value : hist) {
+        const auto value = iter_value.value_iterated_to / output_value_unit_scaling_ratio;
+        const auto percentile = iter_value.percentile_level_iterated_to / 100;
+        const auto total_count = iter_value.total_count_to_this_value;
+
+        // clang-format off
+            out << std::setw(12) << std::defaultfloat << std::setprecision(12) << value
```
Comment (marayl) [inline toolbox/hdr/HdrHistogram.cpp]: Incorrect indentation?

File: toolbox/hdr/HdrHistogram.cpp [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+        out << '\n';
+    }
+
+    const double mean = hist.mean() / output_value_unit_scaling_ratio;
+    const double stddev = hist.stddev();
+    const double max = hist.max() / output_value_unit_scaling_ratio;
+    const double total = hist.get_total_count();
+
+    // clang-format off
+        return out <<
```
Comment (marayl) [inline toolbox/hdr/HdrHistogram.cpp]: Incorrect indentation.

File: toolbox/hdr/HdrHistogram.hpp [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+    HdrHistogram(std::int64_t lowest_value, std::int64_t highest_value,
+                 std::int64_t significant_figures);
+
+    PercentileIterator begin() const;
+    PercentileIterator end() const;
+
+    friend std::ostream& operator<<(std::ostream& out, const HdrHistogram& hist);
+    bool record(std::int64_t value, int count = 1);
+
+    std::int64_t percentile(double percentile);
```
Comment (marayl) [inline toolbox/hdr/HdrHistogram.hpp]: Can this function be `const` and possibly `noexcept`?

File: toolbox/hdr/HdrHistogram.hpp [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+    /// Check whether two values are equivalent (meaning they
+    /// are in the same bucket/range)
+    /// Returns:
+    ///    true if the two values are equivalentv
+    std::int64_t values_are_equivalent(std::int64_t lhs, std::int64_t rhs);
+
+    std::int64_t max() const noexcept;
+    std::int64_t min() const;
```
Comment (marayl) [inline toolbox/hdr/HdrHistogram.hpp]: Add `noexcept` to `min`, `mean`, and `stddev`?

File: toolbox/hdr/HdrHistogram.hpp [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+    std::int64_t get_value_from_sub_bucket(std::int64_t bucket_index,
+                                           std::int64_t sub_bucket_index) const noexcept;
+    std::int64_t get_count_at_index(std::int64_t index) const noexcept;
+    std::int64_t get_target_count_at_percentile(double percentile) const noexcept;
+    std::int64_t counts_index_for(std::int64_t value) const noexcept;
+
+    std::int64_t get_bucket_index(std::int64_t value) const noexcept;
+    std::int64_t get_sub_bucket_index(std::int64_t value, std::int64_t bucket_index) const noexcept;
+
+    std::int64_t get_value_from_index(std::int64_t index) const;
```
Comment (marayl) [inline toolbox/hdr/HdrHistogram.hpp]: Add `noexcept`?

File: toolbox/hdr/HdrHistogramIterator.hpp [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+    std::int64_t count_added_in_this_iter_step{0};
+    std::int64_t total_count_to_this_value{0};
+    std::int64_t total_value_to_this_value{0};
+    double percentile{0.0};
+    double percentile_level_iterated_to{0.0};
+    double int_to_double_conversion_ratio{1.0};
+
+    HdrIterationValue(const AbstractHdrIterator& iterator);
+
+    void set(std::int64_t value);
```
Comment (marayl) [inline toolbox/hdr/HdrHistogramIterator.hpp]: Add `noexcept`?

**Pull request merged**